### PR TITLE
Padding-Free DPO and Some Updates

### PR DESCRIPTION
**Pull request by fabianlim**
This PR superceeds #8. It setups the padding free for `dpo_tune_cache.py`

TODO:
- handle the non-concatenated case for padding free
- update padding-free tests

## Testing with the DataLoader Tester

```python

from datasets import load_dataset, DatasetDict

# create the dataset
load_dataset('HuggingFaceH4/ultrafeedback_binarized')['train_prefs'].select(range(1000)).to_parquet(
    '../datasets/dpo/ultrafeedback_sample.parquet'
)

from scripts.test_dataloading import test_dpo_tune

data = test_dpo_tune(
    model_name_or_path='...',
    train_file='../datasets/dpo/ultrafeedback_sample.parquet',
    padding_free=True,
)

# data[0]
{'input_ids': tensor([[    27,     91,    882,  ...,    828,     13, 100257]]),
 'attention_mask': None,
 'cu_seq_lens_q': tensor([   0,  451, 1013, 1275, 2848, 2987, 3594, 3652, 4276, 4407, 5007, 5109,
         6654, 6777, 7549, 7612, 8218], dtype=torch.int32),
 'cu_seq_lens_k': tensor([   0,  451, 1013, 1275, 2848, 2987, 3594, 3652, 4276, 4407, 5007, 5109,
         6654, 6777, 7549, 7612, 8218], dtype=torch.int32),
 'max_length_q': 1573,
 'max_length_k': 1573,
 'position_ids': tensor([[  0,   1,   2,  ..., 603, 604, 605]]),
 'seq_idx': tensor([[ 0,  0,  0,  ..., 14, 14, 14]], dtype=torch.int32)}
```

### CODE REVIEW

File: open_instruct/dpo_tune_cache.py [hunk 1]
```diff
@@ -602,11 +630,12 @@ def main(args: FlatArguments):
         dataset_args = {}
         if args.train_file is not None:
             data_files["train"] = args.train_file
-        raw_datasets = load_dataset(
-            args.train_file_type,
-            data_files=data_files,
-            **dataset_args,
-        )
+        with accelerator.main_process_first():
```
Comment (Swanand-Kadhe) [inline open_instruct/dpo_tune_cache.py:633]: Quick clarification question: don't we want to add `with accelerator.main_process_first()` on lines 612 and 621?

**Review by Swanand-Kadhe — COMMENTED** [review]
Thanks a lot for the PR. It looks good to me. I just have a small comment.
